src/trajectory.jl

```diff
@@ -517,35 +448,48 @@ Base.:*(d1::Termination, d2::Termination) = Termination(d1.dynamic || d2.dynamic
 isterminated(d::Termination) = d.dynamic || d.numerical
 
 """
-    Termination(s::SliceTS, nt::NUTS, H0::F, H′::F) where {F<:AbstractFloat}
+     $(SIGNATURES)
 
 Check termination of a Hamiltonian trajectory.
 """
-function Termination(s::SliceTS, nt::NUTS, H0::F, H′::F) where {F<:AbstractFloat}
-    return Termination(false, !(s.ℓu < nt.Δ_max + -H′))
+function Termination(s::SliceTS, tc, H0::F, H′::F) where {F<:AbstractFloat}
+    return Termination(false, !(s.ℓu < tc.Δ_max + -H′))
 end
 
 """
-    Termination(s::MultinomialTS, nt::NUTS, H0::F, H′::F) where {F<:AbstractFloat}
+    Termination(s::MultinomialTS, tc, H0::F, H′::F) where {F<:AbstractFloat}
 
 Check termination of a Hamiltonian trajectory.
 """
-function Termination(s::MultinomialTS, nt::NUTS, H0::F, H′::F) where {F<:AbstractFloat}
-    return Termination(false, !(-H0 < nt.Δ_max + -H′))
+function Termination(s::MultinomialTS, tc, H0::F, H′::F) where {F<:AbstractFloat}
+    return Termination(false, !(-H0 < tc.Δ_max + -H′))
 end
 
 """
 A full binary tree trajectory with only necessary leaves and information stored.
 """
-struct BinaryTree{C<:AbstractTerminationCriterion}
+struct BinaryTree{R<:Union{Nothing, AbstractVecOrMat}}
```

Comment:
I am not sure a union type is needed here as Julia allows `undef` value for struct fields. For example:
```julia
julia> struct F2
       a :: Array
       F2() = new()
       end

julia> F2()
F2(#undef)
```
Maybe consider redefine the `BinaryTree` struct, and keep `rho` uninitialised when it is not used (i.e. replacing `Nothing` with `undef`)? This might also offer some performance improvements as it improves type stability.